src/query/phrase_query/exact_phrase_scorer.rs

```diff
@@ -0,0 +1,259 @@
+use std::cmp::Ordering;
+
+use super::phrase_scorer::{PhraseScorer, PostingsWithOffset};
+use crate::docset::{DocSet, TERMINATED};
+use crate::fieldnorm::FieldNormReader;
+use crate::postings::Postings;
+use crate::query::bm25::Bm25Weight;
+use crate::query::phrase_query::phrase_scorer::get_postings_with_offset;
+use crate::query::{Intersection, Scorer};
+use crate::{DocId, Score};
+
+pub struct ExactPhraseScorer<TPostings: Postings> {
+    intersection_docset: Intersection<PostingsWithOffset<TPostings>, PostingsWithOffset<TPostings>>,
+    num_terms: usize,
+    left: Vec<u32>,
+    right: Vec<u32>,
+    phrase_count: u32,
+    fieldnorm_reader: FieldNormReader,
+    similarity_weight: Bm25Weight,
+    scoring_enabled: bool,
```

Comment:
using the name `scoring_enabled` seems weird as it suggests we can keep the scorer and thus use the empty scorer instead.

Reply:
It's an optim used by the `Query` trait https://github.com/saroh/tantivy/blob/1392-phrase-slop-scoring/src/query/query.rs#L47-L48
It allows us to return fast on successful match for our last iteration in the phrase scorer's case. In this case the `BM25Weight` receives a `tf` of 0. So there will be no ranking but docs will still match. Where as `EmptyScorer` will make no docs match apparently.
So we still score but I'd say all documents will have a zero score.

Reply:
added tests to reflect this https://github.com/quickwit-oss/tantivy/pull/1414/files#diff-b46d170c28bc9075958011ee0f1c172d00e2a39169d470250f10151ba7588d3bR190-R236